### PR TITLE
autoinstall: allow cloud source to override iso source

### PR DESCRIPTION
**Pull request by dbungert**
LP: #1948823

### CODE REVIEW

**Review by ogayot — CHANGES_REQUESTED** [review]
I think we should let Subiquity error if the `--autoinstall` path is specified but does not exist. Falling back to a different autoinstall file if not something I'd expect.

File: subiquity/server/server.py [hunk 1]
```diff
@@ -523,19 +530,49 @@ async def wait_for_cloudinit(self):
...
+        if self.opts.autoinstall == "":
+            return None
+
+        locations = (self.base_relative(reload_autoinstall_path),
+                     self.opts.autoinstall,
+                     self.base_relative(cloud_autoinstall_path),
+                     self.base_relative(iso_autoinstall_path))
+
+        for loc in locations:
+            if loc is not None and os.path.exists(loc):
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/server/server.py:560]: I like this idea, but implemented it somewhat differently.  I wanted to raise an exception for the case that someone has gone to the bother to specify the file as an arg but it was then not found.

[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/server/server.py:564]: Yes, that works too. Thanks!

**Review by ogayot — APPROVED** [review]

File: subiquity/server/server.py [hunk 2]
```diff
@@ -523,19 +530,53 @@ async def wait_for_cloudinit(self):
...
+            raise Exception(
+                f'Autoinstall argument {self.opts.autoinstall} not found')
+
+        locations = (self.base_relative(reload_autoinstall_path),
+                     self.opts.autoinstall,
+                     self.base_relative(cloud_autoinstall_path),
+                     self.base_relative(iso_autoinstall_path))
+
+        for loc in locations:
+            if loc is not None and os.path.exists(loc):
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/server/server.py:564]: It's needed for the most common case, where there is no autoinstall arg at all.  Also, `os.path.exists(None)` raises an exception, so I can't let that handle it.

[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/server/server.py:564]: Oh that's right. My bad.

**Comment by mwhudson** [comment]
Aaaa I just had a very sudden and late thought which is that we have documented the fact that an early-command can access and edit the autoinstall config at /autoinstall.yaml: https://ubuntu.com/server/docs/install/autoinstall-reference#early-commands

I guess we can put it there just while early-commands are running?

**Pull request merged**